### ISSUE TITLE
Save CatBoost model Object from Optuna Objective

### ISSUE THREAD

**Issue by ziadzee** [optuna/optuna#2889]
Hello,

I'm currently tuning a CatBoost model using the standard Optuna Objective function approach like so:

```
def objective(trial: optuna.Trial):
    
  ...
  ...
  ...

    
    # Initialsing CatBoost model
    model = cat.CatBoostClassifier(
        loss_function = "CrossEntropy",
        eval_metric = "AUC",
#         task_type = "GPU",
        **params)
    
    model.fit(training_set, early_stopping_rounds = 10)
    
    # Predicting
    prediction = model.predict(validation_set)
    prediction_label = np.rint(prediction)
    
    # Evaluating
    roc_auc = roc_auc_score(y_val, prediction_label)
    f1 = f1_score(y_val, prediction_label)
    recall = recall_score(y_val, prediction_label)
    
    print('ROC AUC Score:', roc_auc)
    print('F1 Score:', f1)
    print('Recall Score:', recall)
    
    return roc_auc, f1, recall
```

However, I'm trying to save the best CatBoost model that was trained and evaluated on my train and validation sets. I want to save the model as an object to then test on my test set. I appreciate Optuna is used for hyperparameter tuning but I wanted to know whether it was possible to just extract the best model directly from my objective function to avoid re-fitting a new model.

I've attempted to follow this [example](https://stackoverflow.com/questions/62144904/python-how-to-retrive-the-best-model-from-optuna-lightgbm-study)  that achieves this using a LightGBM model. Is there a way to achieve this with a CatBoost instead? Thank you

**Comment by nzw0301** [comment]
I think you can use a similar way in the linked page to do so.

**Comment by ziadzee** [comment]
Hello @nzw0301,

I have tried like so:

```
class Objective:
    
    def __init__(self):
        self.best_catboost = None
        self._catboost = None
        
    def __call__(self, trial: optuna.Trial):
        
        features = neverpay.drop(["NPD_EVER"], axis = 1) # X
        target = neverpay["NPD_EVER"] # Y
    
        train_ratio = 0.8
        val_ratio = 0.1
        test_ratio = 0.1

        # Creating train and test sets
        X_train, X_test, y_train, y_test = train_test_split(features,
                                                    target,
                                                    test_size = 1 - train_ratio,
                                                    random_state = 42)

        # Creating validation set
        X_val, X_test, y_val, y_test = train_test_split(X_test,
                                                y_test,
                                                test_size = test_ratio/(test_ratio + val_ratio),
                                                random_state = 42)
    
    
        # Pooling our features and target
    
        training_set = Pool(X_test, y_test)
        validation_set = Pool(X_val, y_val)
        testing_set = Pool(X_train, y_train)
        
        params = {
        
        "iterations" : trial.suggest_int("iterations", 100, 2000, step = 50), # n_estimators (trees)
        "depth" : trial.suggest_int("depth", 2, 10, step = 2),
        "l2_leaf_reg" : trial.suggest_loguniform("l2_leaf_reg", 1e-8, 1.0), #L2 Ridge regularization term on weights
        "learning_rate" : trial.suggest_loguniform("learning_rate", 1e-05, 1e-1),
        "bootstrap_type" : trial.suggest_categorical("bootstap_type", ["Bayesian", "Bernoulli", "MVS"]),
        "random_strength" : trial.suggest_loguniform("random_strength", 1e-8, 1.0),
        "loss_function" : "CrossEntropy",
        "eval_metric" : "AUC",
        "random_seed" : 42,
        "od_type" : "Iter",
        "verbose" : True,
        }
        
        # Initialsing CatBoost model
#         model = cat.CatBoostClassifier(
#         loss_function = "CrossEntropy",
#         eval_metric = "AUC",
#         random_seed = 42,
#         od_type = "Iter",
#         verbose = True,
#         **params)
    
#         model.fit(training_set,
#               early_stopping_rounds = 10)

        model = cat.train(pool = training_set,
                          params = params,
                          early_stopping_rounds = 10)
        
        self._catboost = model
    
        # Predicting
        prediction = model.predict(validation_set)
        prediction_label = np.rint(prediction)
    
        # Evaluating
        roc_auc = roc_auc_score(y_val, prediction_label)
        f1 = f1_score(y_val, prediction_label)
        accuracy = accuracy_score(y_val, prediction_label)
    
        print("Model Best Score:",model.get_best_score())
    
        return roc_auc, f1, accuracy
    
    def callback(self, study, trial):
        if study.best_trials == trial:
            self.best_catboost = self._catboost
```
and when I attempt:

```
objective = Objective()

# Creating the Optuna study for evaluation with 100 trials

study = optuna.create_study(directions = ["maximize", "maximize", "maximize"])


# Optimising study

study.optimize(objective,
               n_trials = 10,
               timeout = 600,
               n_jobs = -1,
               callbacks = [objective.callback])
```

after this, `best_model = objective.best_catboost` returns as a `NoneType`. I believe this is because CatBoost `trian` does not have a `callbacks` parameter?

**Comment by nzw0301** [comment]
I'm afraid that we cannot investigate non-reproducible code.

**Comment by nzw0301** [comment]
Probably, `if study.best_trials == trial` in `callback` always returns `False` since this is the multi-objective optimisation – `study.best_trials` is a list of `FrozenTrial`, not `FrozenTrial`. I would suggest

- Converting the objective function to a single objective function, or
- saving all best trials by replacing the callback logic: 

```python
if study.best_trials[-1] == trial:
    self.best_catboosts.append(self._catboost)  # NOTE: I do not confirm whether or not this works.
```

**Comment by ziadzee** [comment]
Hi, @nzw0301 - thank you for the reply. It seems to work as intended when I convert the objective function to a single objective function, however not have much luck when I attempt with multiple objectives. In this case, I'll just have to re-fit a new CatBoost model with the hyperparameters obtained from my multi-objective function.

**Comment by nzw0301** [comment]
I think we can save catboost model by using [`save_model`](https://catboost.ai/docs/concepts/python-reference_catboost_save_model.html) function rather than re-fitting.
